security-controls/sast.py

```diff
@@ -0,0 +1,311 @@
+"""Update SAST security-controls data."""
+
+import logging
+import os
+import re
+
+import requests
+from google.cloud import firestore
+from requests.auth import HTTPBasicAuth
+
+SECURITY_CONTROLS = os.environ['SC_FIRESTORE_COLLECTION']
+SAST_DETAILS = os.environ['SAST_FIRESTORE_COLLECTION']
+
+SC_PREVIOUS = "security-controls-previous"
+ID = "id"
+SAST = "sast"
+SAST_LINK = "sast_link"
+
+fs = firestore.Client()
+
+CODACY_BASE = "https://app.codacy.com/api/v3"
+CODACY_API_KEY=os.getenv("CODACY_API_KEY").strip()
+CODACY_URL = "https://app.codacy.com/gh/{org}/{repo}/dashboard"
+
+SONARCLOUD_BASE = "https://sonarcloud.io/api"
+SONARCLOUD_API_KEY=os.getenv("SONARCLOUD_API_KEY").strip()
+SONARCLOUD_URL = "https://sonarcloud.io/project/overview?id={project_key}"
+
+GITHUB_GQL_ENDPOINT = "https://api.github.com/graphql"
+GITHUB_TOKEN=os.getenv("GITHUB_TOKEN").strip()
+
+CODACY_HEADERS = {
+    'Accept': 'application/json',
+    'api-token': CODACY_API_KEY
+}
+
+def codacy_org(organization: str) -> str:
+    '''path for codacy org'''
+    return f'{CODACY_BASE}/organizations/gh/{organization}'
+
+
+repo_re = re.compile(r'https://(www.)?github.com/([a-zA-Z0-9\-_]+)/([a-zA-Z0-9\-_]+)')
+RE_GROUP_ORG = 2
+RE_GROUP_REPO = 3
+
+SONAR_ORGS = {
+    "broadinstitute": "dsp-appsec",
+    "DataBiosphere": "broad-databiosphere",
+    "CancerDataAggregator": "cancerdataaggregator"
+}
+
+CODACY_ORGS = [
+    "DataBiosphere",
+    "broadinstitute"
+]
+
+CODACY = "Codacy"
+SONAR = "SonarCloud"
+GITHUB = "GitHub"
+
+RepoKey = tuple[str, str] # organization, repository
+Repo = dict[str, object] # key is source e.g. CODACY, GITHUB; value is repo info from source
+Repos = dict[RepoKey, Repo]
+CodacyOrgData = dict[str, list] # {"org": [ user_suggestion, ... ]}
+
+def get_if(obj: dict, key: str) -> str:
+    '''return obj[value] or empty string if missing'''
+    value = obj.get(key)
+    if value is None:
+        value = ""
+    return value
+
+def repo_list_from_security_controls(repos: Repos):
+    """Initalize repos from Firestore."""
+    sc_collection = fs.collection(SECURITY_CONTROLS)
+    sc_docs = sc_collection.stream()
+    for doc in sc_docs:
+        sc_record = doc.to_dict()
+        repo_url = get_if(sc_record,'github')
+        match = repo_re.match(repo_url)
+        if match is None:
+            logging.warning('SAST controls ignoring %s github="%s"', doc.id, repo_url)
+            continue
+        org = match.group(RE_GROUP_ORG)
+        repo_name = match.group(RE_GROUP_REPO)
+        if org.lower() == 'databiosphere':
+            org = 'DataBiosphere'
+        if org.lower() == 'broadinstitute':
+            org = 'broadinstitute'
+        repo = get_repo(repos, (org, repo_name))
+        sc_record[ID] = doc.id
+        repo[SC_PREVIOUS] = sc_record
+        logging.debug('SAST controls to be updated for %s github="%s"', doc.id, repo_url)
+
+
+def list_github(repos: Repos):
```

Comment:
Why do we need Github metadata for all repos?

Reply:
We're only getting the metadata for repos with SAST running.